### DISCUSSION TITLE
Client.py not receiving video frames from server.py

### DISCUSSION THREAD

**Discussion by prasunnyD** [aiortc/aiortc#624] [Q&A]
I have a server.py file that generates a ball bouncing across a screen and the client.py should stream this ball bouncing. The server.py sends the frames of the ball bouncing to the client.py via VideoStreamTrack. I am using a thread for the ball bouncing infinite loop. I see that the connection between server and client has been made, the frames are being sent from the server, however on the client side nothing happens. For the most part I have been following the videostream-cli and server examples.
### Server.py
```
class BallBouncingVideoStreamTrack(VideoStreamTrack, Thread):
    def __init__(self):
        super().__init__()
        Thread.__init__(self)
        self.daemon = True
        self.start()
        self.queue = Queue(10)
    '''
    In a separate thread the ball bouncing window is launching, where the ball bounces around using an infinite loop.
    '''
    def run(self):
        img = np.zeros((480, 640, 3), dtype='uint8')
        dx, dy = 1, 1
        x, y = 100, 100
        cv2.circle(img, (x, y), 20, (255, 0, 0), -1)
        cv2.imshow('a', img)
        while True:
                    <code for ball bouncing>
                    self.queue.put(img)

    '''
    Returns each frame from the ball bouncing window.
    '''
    async def recv(self):
        img = self.queue.get()
        frame = VideoFrame.from_ndarray(img, format ="bgr24")
        print(frame)
        pts, time_base = await self.next_timestamp()
        frame.pts = pts
        frame.time_base = time_base
        return frame
```
### Client.py
```
class VideoTransformTrack(MediaStreamTrack):
    def __init__(self, track):
        super().__init__()  # don't forget this!
        self.track = track
        print("track id: ",track.id)

    async def recv(self):
        frame = await self.track.recv()
        img = frame.to_ndarray(format="bgr24")
        cv2.imshow("b", img)
        return frame
```

**Comment by rocco-haro** [comment]
Hi @prasunnyD ,

It looks like your example never even starts a negotiation, therefore how can you expect them to communicate when they don't know their counterpart exists?

aiortc has a concept of Offers & Answers. I suggest you carefully study this server code https://github.com/aiortc/aiortc/blob/main/examples/server/server.py#L105 and also study the client code (in javascript). Since you want your client to be purely pythonic, you'll want to reimplement what you observe from the javascript code into python.

Also, since you'll be implementing a client which will have to make HTTP requests to your server, be sure to install and setup `aiohttp_cors` otherwise you'll have issues with completing the offer request.

**Comment by prasunnyD** [comment]
I have established a connection with  server.py sending an offer and client.py sending an answer in response. I didn't include that portion in the question, to minimize the amount of code. 

### server.py
```
async def run(peer_connection, signaling):
    await signaling.connect()
    peer_connection.addTrack(BallBouncingVideoStreamTrack())
    await peer_connection.setLocalDescription(await peer_connection.createOffer())
    await signaling.send(peer_connection.localDescription)
    while True:
        obj = await signaling.receive()
        if isinstance(obj, RTCSessionDescription):
            await peer_connection.setRemoteDescription(obj)
        elif obj is BYE:
            print("Exiting")
            break



if __name__ == "__main__":

    host_name = socket.gethostname()
    host_ip = socket.gethostbyname(host_name)
    print('HOST IP:', host_ip)
    port = 9999

    signaling = TcpSocketSignaling(host_ip, port)
    peer_connection = RTCPeerConnection()
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(run(peer_connection, signaling))
    except KeyboardInterrupt:
        pass
    finally:
        loop.run_until_complete(signaling.close())
        loop.run_until_complete(peer_connection.close())
```

### client.py
```
async def run(pc, signaling):
    @pc.on("track")
    def on_track(track):
        print("Receiving %s" % track.kind)
        if track.kind == "video":
            VideoTransformTrack(track)
            #pc.addTrack(video)

    await signaling.connect()
    while True:
        obj = await signaling.receive()
        if isinstance(obj, RTCSessionDescription):
            await pc.setRemoteDescription(obj)
            if obj.type == "offer":
                # send answer
                await pc.setLocalDescription(await pc.createAnswer())
                await signaling.send(pc.localDescription)
        elif isinstance(obj, RTCIceCandidate):
            await pc.addIceCandidate(obj)
        elif obj is BYE:
            print("Exiting")
            break


if __name__ == "__main__":

    
    host_name = socket.gethostname()
    print(host_name)
    host_ip = socket.gethostbyname(host_name)
    print('HOST IP:', host_ip)
    port = 9999

    signaling = TcpSocketSignaling(host_ip, port)
    peer_connection = RTCPeerConnection()
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(run(peer_connection, signaling))
    except KeyboardInterrupt:
        pass
    finally:
        loop.run_until_complete(signaling.close())
        loop.run_until_complete(peer_connection.close())
```

**Reply by ronenc** [reply]
Hello, have you managed to resolve this issue?

**Comment by ronenc** [comment]
I have the same issue.
I can see that the iceconnectionstate is checking on both ends
![Capture](https://user-images.githubusercontent.com/10868269/154601954-a781fc69-a0dd-450d-981c-7fc32e4178c9.PNG)

sendWebRtcAnswer from  oQY6QQ51TUdTDTfFAAFz
sendWebRtcAnswer to  _7XhcYeWO13lVksuAAFv
DEBUG:aiortc.rtcicetransport:RTCIceTransport(controlled) - new -> checking
DEBUG:aiortc.rtcpeerconnection:RTCPeerConnection() iceConnectionState new -> checking
DEBUG:aiortc.rtcpeerconnection:RTCPeerConnection() connectionState new -> connecting
INFO:aioice.ice:Connection(0) Check CandidatePair(('10.0.0.160', 49562) -> ('10.0.0.160', 61261)) State.FROZEN -> State.WAITING
INFO:aioice.ice:Connection(0) Check CandidatePair(('2604:3d08:d178:3600::8541', 49563) -> ('2604:3d08:d178:3600::8541', 61263)) State.FROZEN -> State.WAITING
INFO:aioice.ice:Connection(0) Check CandidatePair(('2604:3d08:d178:3600:89a7:c338:7796:3e68', 49564) -> ('2604:3d08:d178:3600::8541', 61263)) State.FROZEN -> State.WAITING     
INFO:aioice.ice:Connection(0) Check CandidatePair(('2604:3d08:d178:3600:b57b:89cf:e40f:cfe6', 49565) -> ('2604:3d08:d178:3600::8541', 61263)) State.FROZEN -> State.WAITING     
INFO:aioice.ice:Connection(0) Check CandidatePair(('18.191.196.72', 12461) -> ('10.0.0.160', 61261)) State.FROZEN -> State.WAITING
INFO:aioice.ice:Connection(0) Check CandidatePair(('10.0.0.160', 49562) -> ('18.191.196.72', 17332)) State.FROZEN -> State.WAITING

**Reply by yfz3357** [reply]
hello have you figured this out?
